fix(private-vpn): don't include 99-disable-network-config.cfg if there are only 2 network interfaces

lemeurherve opened the pull request:
Including this cloud-init config seems to break connectivity on Ubuntu Jammy VM like the private.vpn.jenkins.io one.

lemeurherveCB committed:
fix(private-vpn): don't include 99-disable-network-config.cfg if there are only 2 network interfaces

lemeurherveCB committed:
empty file if less than 3 interfaces

dduportal commented on dist/profile/manifests/openvpn.pp:
```diff
@@ -61,15 +61,19 @@
+  # Ensure cloud-init doesn't manage network to ensure the order of eth1 and eth2 if this last one is defined (ie 3 interfaces) (netplan config + netplan apply + systemd, in Ubuntu Bionic)
+  $content = ''
+  if $networks.length > 2 {
+    $content = 'network: {config: disabled}'
```
WDYT putting the `file { ...}` resource block inside the `if  { ... }` block instead? That would ensure that no file is created at all on `private.vpn`.

dduportal reviewed (changes requested):
this PR looks good, but it require to add rspec unit test (you can specify custom hieradata in `hieradata/rspec/...` for instance) to check the 2 cases:

- network more than 2
- networ with only 2

lemeurherveCB committed:
Revert "empty file if less than 3 interfaces"

This reverts commit 776c087f8e177f5c9df6481e892a74f9c3cfca9e.

lemeurherveCB committed:
fix tests

lemeurherveCB committed:
expect no file

lemeurherveCB committed:
correct not exists

dduportal commented on spec/classes/profile/openvpn_spec.rb:
```diff
@@ -29,3 +29,34 @@
 end
+
+
+describe 'profile::openvpn' do
```
```suggestion
describe 'profile::openvpn with 2 interfaces' do
```

dduportal commented on spec/classes/profile/openvpn_spec.rb:
```diff
@@ -29,3 +29,34 @@
+
+  it { expect(subject).to contain_package 'net-tools' }
+
+  it { expect(subject).to contain_file '/etc/cloud/cloud.cfg.d/99-disable-network-config.cfg' }
```
```suggestion
  it { expect(subject).not_to contain_file '/etc/cloud/cloud.cfg.d/99-disable-network-config.cfg' }
```

as per https://github.com/rspec/rspec-expectations/tree/v3.11.0#equivalence

lemeurherve replied:
`.not_to` fails (as `should_not`):

```
Failures:

  1) profile::openvpn with 2 interfaces is expected not to contain File[/etc/cloud/cloud.cfg.d/99-disable-network-config.cfg]
     Failure/Error: it { expect(subject).not_to contain_file '/etc/cloud/cloud.cfg.d/99-disable-network-config.cfg' }
     
     Puppet::ParseErrorWithIssue:
       Could not parse for environment rp_env: This Name has no effect. A value was produced and then forgotten (one or more preceding expressions may have the wrong form) (line: 115, column: 26) on node macbook-pro-de-herve.local
```

lemeurherveCB committed:
apply review suggestions

lemeurherveCB committed:
fix with context

lemeurherveCB committed:
add comment and put in bottom

lemeurherveCB committed:
fix: distinct peered_network_cidrs for eth2

Fails with the following otherwise:

Evaluation Error: Error while evaluating a Resource Statement, Duplicate declaration: Firewall[100 allow routing from 127.0.10.0/24 to 10.0.0.0/16 on ports 80/443] is already declared at (file: ./spec/fixtures/modules/profile/manifests/openvpn.pp, line: 156); cannot redeclare (file: ./spec/fixtures/modules/profile/manifests/openvpn.pp, line: 156) (file: ./spec/fixtures/modules/profile/manifests/openvpn.pp, line: 156, column: 9)

lemeurherveCB committed:
fix

lemeurherve commented on hieradata/rspec/profile_openvpn.yaml:
```diff
@@ -10,6 +10,11 @@ profile::openvpn::networks:
+    name: Another Network
+    network_cidr: 192.168.200.0/24
+    peered_network_cidrs:
+      - 11.0.0.0/16
```
Note: changed from `10.0.0.0/16` to fix this error:

> Evaluation Error: Error while evaluating a Resource Statement, Duplicate declaration: Firewall[100 allow routing from 127.0.10.0/24 to 10.0.0.0/16 on ports 80/443] is already declared at (file: ./spec/fixtures/modules/profile/manifests/openvpn.pp, line: 156); cannot redeclare (file: ./spec/fixtures/modules/profile/manifests/openvpn.pp, line: 156) (file: ./spec/fixtures/modules/profile/manifests/openvpn.pp, line: 156, column: 9)

lemeurherveCB committed:
remove new line

lemeurherveCB committed:
hint private in fixture
